examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,339 @@
+    # tensordict that has shape [N] where N <= B * T.
+    done = td["next", "done"]
+    mask = torch.zeros_like(done)
+    mask[..., 1:, :] = done[..., :-1, :]  # shift by one
+    mask = ~mask.cumsum(-2).bool().squeeze()
+    return td[mask]
+
+
+class AdaptiveKLController:
+    """Adaptive KL Controller as described in Ziegler et al. "Fine-Tuning Language Models from Human Preferences"
+    Reference: Section 2.2 https://arxiv.org/pdf/1909.08593.pdf#page=2
+    Source: https://github.com/openai/lm-human-preferences/blob/master/lm_human_preferences/train_policy.py
+    """
+
+    def __init__(self, init_kl_coef: float, target: float, horizon: int):
+        self.value = init_kl_coef
+        self.target = target
+        self.horizon = horizon
+
+    def update(self, current: float, n_steps: int):
+        """Returns adaptively updated KL coefficient, βₜ₊₁.
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
+        """
+        proportional_error = np.clip(current / self.target - 1, -0.2, 0.2)  # ϵₜ
+        mult = 1 + proportional_error * n_steps / self.horizon
+        self.value *= mult  # βₜ₊₁
+        return self.value
+
+
+def create_reward_estimator(
+    eval_iters, episode_length, reward_model, batch, ctx, logger=None, ref_model=None
+):
+    """Create a function to estimate the reward via sampling.
+
+    This function creates a new function which, given a model and a dataloader, will
+    perform multiple rollouts using the model and data sampled from the dataloader then
+    average the accumulated rewards.
+
+    For debugging purposes, we also generate responses to a fixed prompt so that the
+    quality of the model can be visually assessed during training.
+    """
+    tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+    tokenizer.pad_token = tokenizer.eos_token
+
+    test_rindex = batch.prompt_rindex[0]
+    test_prompt_ids = batch.input_ids[:1, :test_rindex]
+    test_label_ids = batch.input_ids[:1, test_rindex:]
+    generation_config = GenerationConfig(
+        pad_token_id=tokenizer.pad_token_id, max_new_tokens=episode_length
+    )
+    test_prompt = tokenizer.decode(test_prompt_ids[0, :test_rindex].tolist())
+    test_label = tokenizer.decode(
+        test_label_ids[0, test_label_ids[0] != tokenizer.pad_token_id].tolist()
+    )
+    _, test_label_reward = reward_model(
+        input_ids=batch.input_ids[:1], attention_mask=batch.attention_mask[:1]
+    )
+
+    @torch.no_grad()
+    def estimate_reward(model, dataloader):
+        rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+        rewards = torch.zeros(eval_iters)
+        for k in range(eval_iters):
+            batch = next(dataloader)
+            # NOTE: disable kl for evaluation
+            td = rollout_from_model.rollout_from_data(batch, kl_coef=0.0)
+            rewards[k] = td.get(("next", "reward")).sum(dim=1).mean().item()
```

Comment:
why item?

Reply:
to get a scalar instead of a scalar-tensor